### PR TITLE
[Event Hub] No longer create a copy of the array in the message converter for the default and stream case

### PR DESCRIPTION
**Pull request by danielmarbach**
Similar to https://github.com/Azure/azure-sdk-for-net/pull/29946 to make @jsquire happy

# Contributing to the Azure SDK

Please see our [CONTRIBUTING.md](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md) if you are not familiar with contributing to this repository or have questions.

For specific information about pull request etiquette and best practices, see [this section](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md#pull-request-etiquette-and-best-practices).

### CODE REVIEW

**Comment by danielmarbach** [comment]
The failing test is passing for me locally. Could this be a race and a problem of the closure capturing of the bool in these tests?

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Amqp/AmqpMessageConverter.cs
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs/src/Amqp/AmqpMessageConverter.cs:1228]:
```suggestion
                    if (!memStreamSource.TryGetBuffer(out ArraySegment<byte> segment))
                    {
                        using var memStreamCopy = new MemoryStream((int)(memStreamSource.Length - memStreamSource.Position));
                        memStreamSource.CopyTo(memStreamCopy, StreamBufferSizeInBytes);
                        
                        segment = new ArraySegment<byte>(memStreamCopy.ToArray());
                    }
                    return segment;
                }
```

If you're feeling like trying something, I'm very curious if this would work.   From the docs that you shared on the SB changes, I **_think_** that it should, since we know its a memory stream and has an internal buffer.    If we pass the live suite with this, I'm inclined to take it and push it through our stress tests for a few days to see what happens.

Thoughs?

**Comment by jsquire** [comment]
> closure capturing of the bool in these tests?

It's possible.   That test was not easy to get the order of operations and timing right.   I thought that I had it stabilized, but I may need to try a volatile read in the closure.

File: sdk/eventhub/Azure.Messaging.EventHubs/CHANGELOG.md
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs/CHANGELOG.md:12]: Don't forget to add this sucker before the "Features Added" header:

```markdown
### Acknowledgments

Thank you to our developer community members who helped to make the Event Hubs client libraries better with their contributions to this release:

- Daniel Marbach _([GitHub](https://github.com/danielmarbach))_
```

It's not much but we can at least highlight your efforts rather than burying them inline.  😄

**Review by jsquire — COMMENTED** [review]
LGTM;  holding off clicking the approval only to hear your thoughts on the proposed experiment.   
_(and a live test run once we decide that we're done)_

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Amqp/AmqpMessageConverter.cs
Comment (danielmarbach) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/Amqp/AmqpMessageConverter.cs:1228]: I think the problem is the following: I assume the most likely source here is AmqpMessage. AmqpMessage uses [`MemoryStream(Byte[], Int32, Int32)`](https://docs.microsoft.com/en-us/dotnet/api/system.io.memorystream.-ctor?view=net-6.0#system-io-memorystream-ctor(system-byte()-system-int32-system-int32)) which means `TryGetBuffer` will always return `false`and then we would fall back creating a memory stream with a predefined capacity (of which `TryGetBuffer` would return `true`) only then to write to it and then copy the internal buffer again into the segment because we are calling `ToArray` (which is the unnecessary double allocation this PR gets rid off). 

So what would we gain from this experiment under that promise? Or am I misunderstanding that the sources are most likely AmqpMessages?

What would help is to do a double `TryGetBuffer`. First on the passed stream, then if it returns false allocate the memory stream with the length, copy stuff into it and then `TryGetBuffer` to return it to avoid the `ToArray`. But then again how much value would that add in case we are always getting MemoryStreams managed by AmqpMessage and their internal buffering (still assuming that is indeed the case)?

Comment (jsquire) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/Amqp/AmqpMessageConverter.cs:1228]: That's where my disconnect is;  the "only works with some constructors".   I think the majority of the time, we'll hit the `BufferListStream` case anyway, coming from `AmqpMessage`, just looking over the [ToStream](https://github.com/Azure/azure-amqp/blob/hotfix/Microsoft.Azure.Amqp/Amqp/AmqpMessage.cs) implementations.    The only other case where we'll hit this method is when someone puts a `Stream` inside the application properties - which is a very rare occurrence.

I agree, not worth the experiment and though the double try-get would guard, unlikely worth the complexity.

**Review by jsquire — APPROVED** [review]
LGTM; when you're ready to marge, give a yell.

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Amqp/AmqpMessageConverter.cs
Comment (danielmarbach) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/Amqp/AmqpMessageConverter.cs:1228]: How about we merge this as is and I spend some more time after my vacation to think about these scenarios in more depth and then send follow up PRs?

**Comment by danielmarbach** [comment]
From my standpoint this is already a nice improvement and as such valuable and good to go

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Amqp/AmqpMessageConverter.cs
Comment (jsquire) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/Amqp/AmqpMessageConverter.cs:1228]: Agreed; not sure how much more we can optimize and we're hitting the majority cases.   Thank you, once again, for all of your work on this!

**Pull request merged**
